File: adf_core_python/core/agent/info/agent_info.py
from __future__ import annotations

from time import time
from typing import TYPE_CHECKING, Any

from rcrs_core.commands.Command import Command
from rcrs_core.entities.civilian import Civilian
from rcrs_core.entities.entity import Entity
from rcrs_core.entities.human import Human
from rcrs_core.worldmodel.changeSet import ChangeSet
from rcrs_core.worldmodel.entityID import EntityID
from rcrs_core.worldmodel.worldmodel import WorldModel

from adf_core_python.core.agent.action.action import Action

if TYPE_CHECKING:
    from adf_core_python.core.agent.agent import Agent


class AgentInfo:
    # TODO: Replace Any with the actual type
    def __init__(self, agent: Agent, world_model: WorldModel):
        self._agent: Agent = agent
        self._world_model: WorldModel = world_model
        self._time: int = 0
        self._action_history: dict[int, Action] = {}
        self._heard_commands: list[Any] = []
        self._change_set: ChangeSet = ChangeSet()
        self._start_think_time: float = 0.0

    def set_time(self, time: int) -> None:
        """
        Set the current time of the agent

        Parameters
        ----------
        time : int
            Current time
        """
        self._time = time

    def get_time(self) -> int:
        """
        Get the current time of the agent

        Returns
        -------
        int
            Current time
        """
        return self._time

    def set_heard_commands(self, heard_commands: list[Command]) -> None:
        """
        Set the heard commands

        Parameters
        ----------
        heard_commands : list[Command]
            Heard commands
        """
        self._heard_commands = heard_commands

    def get_heard_commands(self) -> list[Command]:
        """
        Get the heard commands

        Returns
        -------
        list[Command]
            Heard commands
        """
        return self._heard_commands

    def get_entity_id(self) -> EntityID:
        """
        Get the entity ID of the agent

        Returns
        -------
        EntityID
            Entity ID of the agent
        """
        # TODO: Agent class should return EntityID instead of EntityID | None
        return self._agent.get_entity_id()

    def get_myself(self) -> Entity:
        """
        Get the entity of the agent

        Returns
        -------
        Entity
            Entity of the agent
        """
        return self._world_model.get_entity(self.get_entity_id())

    def get_position_entity_id(self) -> EntityID:
        """
        Get the position entity ID of the agent

        Returns
        -------
        EntityID
            Position entity ID of the agent
        """
        entity = self._world_model.get_entity(self.get_entity_id())
        if isinstance(entity, Human):
            return entity.get_position()
        else:
            return entity.get_id()

    def set_change_set(self, change_set: ChangeSet) -> None:
        """
        Set the change set

        Parameters
        ----------
        change_set : ChangeSet
            Change set
        """
        self._change_set = change_set

    def get_change_set(self) -> ChangeSet:
        """
        Get the change set

        Returns
        -------
        ChangeSet
            Change set
        """
        return self._change_set

    def some_one_on_board(self) -> Human | None:
        """
        Get the human if someone is on board

        Returns
        -------
        Human | None
            Human if someone is on board, None otherwise
        """
        entity_id: EntityID = self.get_entity_id()
        for entity in self._world_model.get_entities():
            if isinstance(entity, Civilian):
                if entity.get_position() == entity_id:
                    return entity
        return None

    def get_executed_action(self, time: int) -> Action | None:
        """
        Get the executed action at the given time

        Parameters
        ----------
        time : int
            Time
        """
        return self._action_history.get(time)

    def set_executed_action(self, time: int, action: Action) -> None:
        """
        Set the executed action at the given time

        Parameters
        ----------
        time : int
            Time
        action : Action
            Executed action
        """
        self._action_history[time] = action

    def record_think_start_time(self) -> None:
        """
        Record the start time of thinking
        """
        self._start_think_time = time()

    def get_think_time(self) -> float:
        """
        Get the time taken for thinking

        Returns
        -------
        float
            Time taken for thinking
        """
        return time() - self._start_think_time

File: java/lib/src/main/java/adf_core_python/agent/info/AgentInfo.java
package adf_core_python.agent.info;

import adf.core.agent.action.Action;
import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;
import rescuecore2.messages.Command;
import rescuecore2.standard.entities.*;
import rescuecore2.worldmodel.ChangeSet;
import rescuecore2.worldmodel.EntityID;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class AgentInfo {
    private final EntityID entityID;
    private final StandardWorldModel worldModel;
    private int time;
    private ChangeSet changed;
    private Collection<Command> heard;
    private long thinkStartTime;

    private Map<Integer, Action> actionHistory;

    public AgentInfo(@Nonnull EntityID entityID, @Nonnull StandardWorldModel worldModel) {
        this.entityID = entityID;
        this.worldModel = worldModel;
        this.time = 0;
        this.actionHistory = new HashMap<>();
        recordThinkStartTime();
    }

    public int getTime() {
        return this.time;
    }

    public void setTime(int time) {
        this.time = time;
    }

    @Nullable
    public Collection<Command> getHeard() {
        return this.heard;
    }

    public void setHeard(Collection<Command> heard) {
        this.heard = heard;
    }

    @Nonnull
    public EntityID getID() {
        return this.entityID;
    }

    public StandardEntity me() {
        return this.worldModel.getEntity(this.getID());
    }

    public double getX() {
        return this.worldModel.getEntity(this.entityID).getLocation(this.worldModel).first();
    }

    public double getY() {
        return this.worldModel.getEntity(this.entityID).getLocation(this.worldModel).second();
    }

    public EntityID getPosition() {
        StandardEntity entity = this.worldModel.getEntity(this.getID());
        return entity instanceof Human ? ((Human) entity).getPosition() : entity.getID();
    }

    @Nonnull
    public Area getPositionArea() {
        return (Area) this.worldModel.getEntity(this.getPosition());
    }

    @Nullable
    public ChangeSet getChanged() {
        return this.changed;
    }

    public void setChanged(ChangeSet changed) {
        this.changed = changed;
    }

    @Nullable
    public Human someoneOnBoard() {

        for (StandardEntity next : this.worldModel
                .getEntitiesOfType(StandardEntityURN.CIVILIAN)) {
            Human human = (Human) next;
            if (human.getPosition().equals(this.entityID)) {
                return human;
            }
        }

        return null;
    }

    @Nullable
    public Action getExecutedAction(int time) {
        if (time > 0)
            return this.actionHistory.get(time);
        return this.actionHistory.get(this.getTime() + time);
    }


    public void setExecutedAction(int time, @Nullable Action action) {
        this.actionHistory.put(time > 0 ? time : this.getTime() + time, action);
    }

    public void recordThinkStartTime() {
        this.thinkStartTime = System.currentTimeMillis();
    }

    public long getThinkTimeMillis() {
        return (System.currentTimeMillis() - this.thinkStartTime);
    }
}
